pkg/cosign/verify.go

```diff
@@ -1130,155 +1137,80 @@ func comparePublicKey(bundleBody string, sig oci.Signature, co *CheckOpts) error
-		if err != nil {
-			return "", "", err
-		}
-		err = json.Unmarshal(specMarshal, &rekordObj)
-		if err != nil {
-			return "", "", err
-		}
-		return *rekordObj.Data.Hash.Algorithm, *rekordObj.Data.Hash.Value, nil
-	}
+	return rekor_types.UnmarshalEntry(pe)
+}
 
-	// Try hashedRekordObj
-	err = json.Unmarshal(bodyDecoded, &hrekord)
-	if err != nil {
-		return "", "", err
-	}
-	specMarshal, err := json.Marshal(hrekord.Spec)
+func bundleHash(bundleBody, _ string) (string, string, error) {
+	ei, err := extractEntryImpl(bundleBody)
 	if err != nil {
 		return "", "", err
 	}
-	err = json.Unmarshal(specMarshal, &hrekordObj)
-	if err != nil {
-		return "", "", err
+
+	switch entry := ei.(type) {
+	case *dsse_v001.V001Entry:
+		return *entry.DSSEObj.EnvelopeHash.Algorithm, *entry.DSSEObj.EnvelopeHash.Value, nil
+	case *hashedrekord_v001.V001Entry:
+		return *entry.HashedRekordObj.Data.Hash.Algorithm, *entry.HashedRekordObj.Data.Hash.Value, nil
+	case *intoto_v001.V001Entry:
+		return *entry.IntotoObj.Content.Hash.Algorithm, *entry.IntotoObj.Content.Hash.Value, nil
+	case *intoto_v002.V002Entry:
+		return *entry.IntotoObj.Content.Hash.Algorithm, *entry.IntotoObj.Content.Hash.Value, nil
+	case *rekord_v001.V001Entry:
+		return *entry.RekordObj.Data.Hash.Algorithm, *entry.RekordObj.Data.Hash.Value, nil
+	default:
+		return "", "", errors.New("unsupported type")
 	}
-	return *hrekordObj.Data.Hash.Algorithm, *hrekordObj.Data.Hash.Value, nil
 }
 
 // bundleSig extracts the signature from the rekor bundle body
 func bundleSig(bundleBody string) (string, error) {
-	var rekord models.Rekord
-	var hrekord models.Hashedrekord
-	var rekordObj models.RekordV001Schema
-	var hrekordObj models.HashedrekordV001Schema
-
-	bodyDecoded, err := base64.StdEncoding.DecodeString(bundleBody)
-	if err != nil {
-		return "", fmt.Errorf("decoding bundleBody: %w", err)
-	}
-
-	// Try Rekord
-	if err := json.Unmarshal(bodyDecoded, &rekord); err == nil {
-		specMarshal, err := json.Marshal(rekord.Spec)
-		if err != nil {
-			return "", err
-		}
-		if err := json.Unmarshal(specMarshal, &rekordObj); err != nil {
-			return "", err
-		}
-		return rekordObj.Signature.Content.String(), nil
-	}
-
-	// Try hashedRekordObj
-	if err := json.Unmarshal(bodyDecoded, &hrekord); err != nil {
-		return "", err
-	}
-	specMarshal, err := json.Marshal(hrekord.Spec)
+	ei, err := extractEntryImpl(bundleBody)
 	if err != nil {
 		return "", err
 	}
-	if err := json.Unmarshal(specMarshal, &hrekordObj); err != nil {
-		return "", err
+
+	switch entry := ei.(type) {
+	case *dsse_v001.V001Entry:
+		// TODO: this could have multiple signatures but we're only returning the first here
+		return *entry.DSSEObj.Signatures[0].Signature, nil
+	case *hashedrekord_v001.V001Entry:
+		return entry.HashedRekordObj.Signature.Content.String(), nil
+	case *rekord_v001.V001Entry:
+		return entry.RekordObj.Signature.Content.String(), nil
+	default:
+		return "", errors.New("unsupported type")
```

Comment:
What about intoto types? I see we didn't try to extract signatures previously, but isn't that a gap?

Reply:
in `compareSigs` if an actual attestation is stored in OCI, the signature from the DSSE envelope is not stored as a separate file (as it would be with other types) so the verification is done elsewhere.